### ISSUE TITLE
Unhandled server exception

### ISSUE THREAD

**Issue by mconlow** [G-Two/subarulink#10]
Hope you don't mind if I open this as an issue. I've been having a lot of problems using the MySubaru app and the command line script tonight. It has actually crossed my mind that I'm being rate limited by their server. Anyhow, one of the errors I'm getting for lock/unlock commands is this:

2020-07-20 22:38:59,091 - subarulink.connection - DEBUG - GET: https://mobileapi.prod.subarucs.com/g2v15/validateSession.json
Traceback (most recent call last):
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/connection.py", line 229, in __open
    resp = await getattr(self.websession, method)(url, headers=headers, params=params, data=data, json=json)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/aiohttp/client.py", line 504, in _request
    await resp.start(conn)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/aiohttp/client_reqrep.py", line 847, in start
    message, payload = await self._protocol.read()  # type: ignore  # noqa
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/aiohttp/streams.py", line 591, in read
    await self._waiter
aiohttp.client_exceptions.ServerDisconnectedError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/local_user/.pyenv/versions/3.7.3/bin/subarulink", line 10, in <module>
    sys.exit(main())
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/app/cli.py", line 554, in main
    LOOP.run_until_complete(cli.single_command(args.command, args.vin))
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
    return future.result()
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/app/cli.py", line 422, in single_command
    success = await self._ctrl.unlock(self._current_vin)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 198, in unlock
    resp = await self._actuate(vin, "unlock", data=form_data)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 305, in _actuate
    return await self._remote_command(vin, cmd, data=form_data)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 284, in _remote_command
    await self._connection.validate_session(vin)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/connection.py", line 64, in validate_session
    resp = await self.__open("/validateSession.json", "get")
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/connection.py", line 238, in __open
    raise SubaruException("aiohttp.ClientConnectionError")
subarulink.exceptions.SubaruException
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x10a2e2fd0>

**Comment by mconlow** [comment]
Then a couple seconds later with the same command (unlock) it appears to get through that step but then errors a couple requests down this time suggesting it got a 403 from the server:

```
2020-07-20 22:46:06,711 - subarulink.connection - DEBUG - GET: https://mobileapi.prod.subarucs.com/g2v15/service/g2/remoteService/status.json
2020-07-20 22:46:07,043 - subarulink.controller - DEBUG - {'data': {'errorDescription': None,
          'errorLabel': '403-soa-unableToParseResponseBody'},
 'dataName': 'errorResponse',
 'errorCode': '403-soa-unableToParseResponseBody',
 'success': False}
Traceback (most recent call last):
  File "/Users/local_user/.pyenv/versions/3.7.3/bin/subarulink", line 10, in <module>
    sys.exit(main())
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/app/cli.py", line 554, in main
    LOOP.run_until_complete(cli.single_command(args.command, args.vin))
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
    return future.result()
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/app/cli.py", line 422, in single_command
    success = await self._ctrl.unlock(self._current_vin)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 198, in unlock
    resp = await self._actuate(vin, "unlock", data=form_data)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 305, in _actuate
    return await self._remote_command(vin, cmd, data=form_data)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 295, in _remote_command
    js_resp = await self._wait_request_status(req_id, api_gen, poll_url)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 348, in _wait_request_status
    if js_resp["data"]["success"]:
KeyError: 'success'
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x107248048>
```

**Comment by G-Two** [comment]
Thanks, I've been seeing the same 403 error in my Home Assistant logs once every few days and haven't gotten a good stack trace.  This definitely needs to be handled in the controller.  I've been meaning to do more rigorous checking of what comes back from the server in Controller._wait_request_status().  I'm not sure of the root cause, but I suspect it is something on the server side since the application received a service request ID for the remote command and was in the wait loop for command completion.  Do you have any log entries just prior to the 403 error?  I'm curious if any valid wait responses were received.

As for the ServerDisconnectedError, I've never received one of these.  It does seem like it could be rate limiting from the Subaru servers.  How frequently were you sending commands at that time?  I'll add handling for this exception.

Also, out of curiosity, what year/model vehicle are you testing on?   Aside from the errors, have any of the remote functions worked for you?

**Comment by mconlow** [comment]
It's a 2020 Ascent. I'll check later if I still have more logging. I got a slew of success messages overnight so maybe they were just queued up on their side? I was hammering it doing testing so it could also be rate limiting. The app itself was also erroring, sometimes not letting me in at all. So it could also be API instability or downtime...

I've been mostly trying the "lock" command which has definitely worked most of the time before last night. I was planning to move on to remote-start after that.

My plan is to make a HA "restful switch"  for the lock and remote start commands. (I have a little Flask server I can use kind of as a proxy). It seems to me that HA issues GET requests to check the status of the endpoint quite frequently. Combine that with a slow API from Subaru and not actually being able to receive the actual lock status, and I may just fake the GET endpoint.

**Comment by stboch** [comment]
The rate limiting can't be that low for just status queries the mysubaru app checks for status every 2 seconds after a command is issued until the status goes to finished.

**Comment by mconlow** [comment]
Here's what's happening directly above the status.json call and the 403 error:

```
2020-07-20 22:46:03,097 - subarulink.connection - DEBUG - POST: https://mobileapi.prod.subarucs.com/g2v15/service/g2/unlock/execute.json
2020-07-20 22:46:04,151 - subarulink.controller - DEBUG - {'data': {'cancelled': False,
          'errorCode': None,
          'remoteServiceState': 'started',
          'remoteServiceType': 'unlock',
          'result': None,
          'serviceRequestId': '****_1595299564071_19_@NGTP',
          'subState': None,
          'success': False,
          'updateTime': None,
          'vin': '*****'},
 'dataName': 'remoteServiceStatus',
 'errorCode': None,
 'success': True}
2020-07-20 22:46:04,151 - subarulink.controller - DEBUG - Polling for remote service request completion: serviceRequestId=******_1595299564071_19_@NGTP
2020-07-20 22:46:04,151 - subarulink.connection - DEBUG - GET: https://mobileapi.prod.subarucs.com/g2v15/service/g2/remoteService/status.json
2020-07-20 22:46:04,707 - subarulink.controller - DEBUG - {'data': {'cancelled': False,
          'errorCode': None,
          'remoteServiceState': 'started',
          'remoteServiceType': 'unlock',
          'result': None,
          'serviceRequestId': '*****_1595299564071_19_@NGTP',
          'subState': None,
          'success': False,
          'updateTime': 1595299564000,
          'vin': '*****'},
 'dataName': 'remoteServiceStatus',
 'errorCode': None,
 'success': True}

```

**Comment by mconlow** [comment]
I just did another one, got the same 403 and an error from the Controller, but I got a push notification that the lock command had succeeded.

```
2020-07-21 12:37:44,277 - subarulink.connection - DEBUG - POST: https://mobileapi.prod.subarucs.com/g2v15/service/g2/lock/execute.json
2020-07-21 12:37:44,707 - subarulink.controller - DEBUG - {'data': {'cancelled': False,
          'errorCode': None,
          'remoteServiceState': 'started',
          'remoteServiceType': 'lock',
          'result': None,
          'serviceRequestId': '******_1595349496614_20_@NGTP',
          'subState': None,
          'success': False,
          'updateTime': None,
          'vin': '******'},
 'dataName': 'remoteServiceStatus',
 'errorCode': None,
 'success': True}
2020-07-21 12:37:44,708 - subarulink.controller - DEBUG - Polling for remote service request completion: serviceRequestId=******_1595349496614_20_@NGTP
2020-07-21 12:37:44,708 - subarulink.connection - DEBUG - GET: https://mobileapi.prod.subarucs.com/g2v15/service/g2/remoteService/status.json
2020-07-21 12:37:44,811 - subarulink.controller - DEBUG - {'data': {'cancelled': False,
          'errorCode': None,
          'remoteServiceState': 'started',
          'remoteServiceType': 'lock',
          'result': None,
          'serviceRequestId': '******_1595349496614_20_@NGTP',
          'subState': None,
          'success': False,
          'updateTime': 1595349496000,
          'vin': '******'},
 'dataName': 'remoteServiceStatus',
 'errorCode': None,
 'success': True}
2020-07-21 12:37:46,812 - subarulink.connection - DEBUG - GET: https://mobileapi.prod.subarucs.com/g2v15/service/g2/remoteService/status.json
2020-07-21 12:37:46,895 - subarulink.controller - DEBUG - {'data': {'errorDescription': None,
          'errorLabel': '403-soa-unableToParseResponseBody'},
 'dataName': 'errorResponse',
 'errorCode': '403-soa-unableToParseResponseBody',
 'success': False}
Traceback (most recent call last):
  File "/Users/local_user/.pyenv/versions/3.7.3/bin/subarulink", line 10, in <module>
    sys.exit(main())
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/app/cli.py", line 554, in main
    LOOP.run_until_complete(cli.single_command(args.command, args.vin))
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
    return future.result()
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/app/cli.py", line 419, in single_command
    success = await self._ctrl.lock(self._current_vin)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 188, in lock
    resp = await self._actuate(vin, "lock", data=form_data)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 305, in _actuate
    return await self._remote_command(vin, cmd, data=form_data)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 295, in _remote_command
    js_resp = await self._wait_request_status(req_id, api_gen, poll_url)
  File "/Users/local_user/.pyenv/versions/3.7.3/lib/python3.7/site-packages/subarulink/controller.py", line 348, in _wait_request_status
    if js_resp["data"]["success"]:
KeyError: 'success'
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x10b14dfd0>
Unclosed connector
connections: ['[(<aiohttp.client_proto.ResponseHandler object at 0x10b118ee8>, 10.210713795)]']
connector: <aiohttp.connector.TCPConnector object at 0x10b14df98>
```

**Comment by stboch** [comment]
@G-Two 
The issue might be in _wait_request_status
You are chhecking for data.success but in my review this does not indicate job state, what you need to check for is if js_resp["data"]["remoteServiceState"] == "finished"
because one a request can fail but is complete...

**Comment by G-Two** [comment]
@mconlow thanks for the log entries.  Good to see that the remote command is making it to the vehicle.  Looks like I just need to handle the occasional '403' gracefully and try again to get the completion message so the user will get the feedback.
@stboch I agree, I've collected a few log entries of failed but complete remote commands and will use them to emit more useful SubaruExceptions that provide more detail about various types of failures.

**Comment by G-Two** [comment]
I pushed a new commit that should address both of these unhandled exceptions.  _wait_request_status() now checks the correct JSON fields to determine remote command completion and success.  If a 403 (or other error) is received, it is just logged and tries to continue.  Since ServerDisconnectedError is so low level, I just pass it on as a SubaruException and let the application deal with it however it wants.  I hopefully didn't introduce any new bugs with these changes.

**Comment by G-Two** [comment]
I believe this issue has been fixed as of 0.3.9, maybe earlier.
